### PR TITLE
Split XDMA and interrupt setup/teardown

### PR DESCRIPTION
**Pull request by iomartin**
PR #23 requires setting up interrupts for the eBPF engines. However, the XDMA and interrupt setup/teardown were tied together, so adding the eBPF interrupts would have been awkward/hacky.

This PR cleans that up by moving all the interrupt setup/teardown out of XDMA functions and files into hermes_mod.c.

### CODE REVIEW

File: src/driver/libxdma.c [hunk 1]
```diff
@@ -1407,6 +1402,16 @@ static int irq_setup(struct xdma_dev *xdev)
...
 
+static void irq_teardown(struct xdma_dev *xdev)
+{
+	irq_msix_channel_teardown(xdev);
+}
+
+static int irq_setup(struct xdma_dev *xdev)
+{
+	return xdma_irq_setup(xdev);
+}
```
Comment (lsgunth) [inline src/driver/libxdma.c]: I don't understand why we need these tiny wrappers that just confuse the names? Can't you just rename the call sites?

File: src/driver/hermes_mod.c [hunk 2]
```diff
@@ -55,12 +55,102 @@ static const struct pci_device_id pci_ids[] = {
...
 
+static void irq_teardown(struct xdma_dev *xdev)
+{
+	irq_msix_channel_teardown(xdev);
+}
+
+static int irq_setup(struct xdma_dev *xdev)
+{
+	return xdma_irq_setup(xdev);
+}
```
Comment (lsgunth) [inline src/driver/hermes_mod.c:66]: Same here... What's the point of another set of wrappers to rename the functions?

File: src/driver/hermes_mod.c [hunk 3]
```diff
@@ -55,12 +55,102 @@ static const struct pci_device_id pci_ids[] = {
...
+
+/*
+ * code to detect if MSI/MSI-X capability exists is derived
+ * from linux/pci/msi.c - pci_msi_check_device
+ */
+
+#ifndef arch_msi_check_device
+static int arch_msi_check_device(struct pci_dev *dev, int nvec, int type)
```
Comment (lsgunth) [inline src/driver/hermes_mod.c:74]: This is highly questionable.... Why is the IRQ handling of this device so complicated and special?

File: src/driver/hermes_mod.c [hunk 4]
```diff
@@ -55,12 +55,102 @@ static const struct pci_device_id pci_ids[] = {
...
+
+	if (unlikely(!xdev || !pdev)) {
+		pr_err("xdev 0x%p, pdev 0x%p.\n", xdev, pdev);
+		return -EINVAL;
+	}
+
+	if (msi_msix_capable(pdev, PCI_CAP_ID_MSIX)) {
+		int req_nvec = xdev->c2h_channel_max + xdev->h2c_channel_max;
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)
```
Comment (lsgunth) [inline src/driver/hermes_mod.c:122]: Do we really need to support kernels that are that old?

File: src/driver/hermes_mod.c [hunk 5]
```diff
@@ -55,12 +55,102 @@ static const struct pci_device_id pci_ids[] = {
...
+{
+	struct pci_dev *pdev = xdev->pdev;
+	int rv = 0;
+
+	if (unlikely(!xdev || !pdev)) {
+		pr_err("xdev 0x%p, pdev 0x%p.\n", xdev, pdev);
+		return -EINVAL;
+	}
+
+	if (msi_msix_capable(pdev, PCI_CAP_ID_MSIX)) {
```
Comment (lsgunth) [inline src/driver/hermes_mod.c:119]: Seems like this janky msi_msix_capable() function is better handled by pci_alloc_irq_vectors() or pci_enable_msix(), which will fail if it can't deliver MSIX vectors....

**Review by lsgunth — COMMENTED** [review]
Looks like another xdma mess that's really hard to follow and way more complicated than it needs to be....

File: src/driver/hermes_mod.c [hunk 6]
```diff
@@ -55,12 +55,102 @@ static const struct pci_device_id pci_ids[] = {
...
+		int req_nvec = xdev->c2h_channel_max + xdev->h2c_channel_max;
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,12,0)
+		pr_debug("Enabling MSI-X\n");
+		rv = pci_alloc_irq_vectors(pdev, req_nvec, req_nvec,
```
Comment (lsgunth) [inline src/driver/hermes_mod.c:124]: I'm also not sure I'm following exactly what's going on with all these moves and renames (bit of a rats nest).... But pci_alloc_irq_vectors() really should only be called by the PCI driver exactly once. It's not clear if that is still true here...

Comment (iomartin) [reply src/driver/hermes_mod.c:124]: Yeah, the main motivation for this PR was that the call to pci_alloc_irq_vectors() was deep inside the xdma init and we needed to add the extra vectors for eBPF. The new code calls it only once.

File: src/driver/libxdma.c (hunk 1 above)
Comment (iomartin) [reply src/driver/libxdma.c]: Future patches move irq_setup() out of libxdma.c and add the eBPF irq setup to it.

However, a function on libxdma.c (xdma_device_online(), which is called by slot_reset()) still needs to call irq_setup(), but without the future eBPF stuff.

File: src/driver/hermes_mod.c (hunk 3 above)
Comment (iomartin) [reply src/driver/hermes_mod.c:74]: Hmm this might be a leftover of their support for ancient kernels. Or not. But anyway, you're right, this and msi_msix_capable() aren't required

**Comment by iomartin** [comment]
I pushed a new version with some other cleanups and reorganized the code a bit more.

- Commits 0-3 clean up the code with no functional changes.
- Commits 4-5 rename the irq setup/teardown functions to make it clear they are doing XDMA stuff only and wrap them around irq_setup/teardown. When we add eBPF interrupts, we'll add them to irq_setup()/irq_teardown().
- Commit 6 notices that calls to irq_setup (irq_teardown) are always followed (preceeded) by a couple of function calls, so move them into irq_setup (irq_teardown).
- Commit 7 splits the XDMA and interrupt setup/teardown
- Commit 8 then cleans the code a bit further

File: src/driver/hermes_mod.c [hunk 7]
```diff
@@ -55,24 +55,25 @@ static const struct pci_device_id pci_ids[] = {
 };
 MODULE_DEVICE_TABLE(pci, pci_ids);
 
-static void irq_teardown(struct xdma_dev *xdev)
+static void irq_teardown(struct hermes_pci_dev *hpdev)
```
Comment (lsgunth) [inline src/driver/hermes_mod.c:58]: I got to the end of the patch set and I still don't see these being used for anything.

Why not open code these now, then create a new helper when you actually have something to add to it?

Besides that this seems like some decent cleanup.

Comment (iomartin) [reply src/driver/hermes_mod.c:58]: Sure, I can move some of these commits to #23 where they'll be used.

**Review by lsgunth — APPROVED** [review]
Looks good now, thanks
